cmd/osv-scanner/main_test.go

```diff
@@ -0,0 +1,242 @@
+//nolint:testpackage // main cannot be accessed directly, so cannot use main_test
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strings"
+	"testing"
+)
+
+func dedent(t *testing.T, str string) string {
+	t.Helper()
+
+	// 0. replace all tabs with spaces
+	str = strings.ReplaceAll(str, "\t", "  ")
+
+	// 1. remove trailing whitespace
+	re := regexp.MustCompile(`\r?\n([\t ]*)$`)
+	str = re.ReplaceAllString(str, "")
+
+	// 2. if any of the lines are not indented, return as we're already dedent-ed
+	re = regexp.MustCompile(`(^|\r?\n)[^\t \n]`)
+	if re.MatchString(str) {
+		return str
+	}
+
+	// 3. find all line breaks to determine the highest common indentation level
+	re = regexp.MustCompile(`\n[\t ]+`)
+	matches := re.FindAllString(str, -1)
+
+	// 4. remove the common indentation from all strings
+	if matches != nil {
+		size := len(matches[0]) - 1
+
+		for _, match := range matches {
+			if len(match)-1 < size {
+				size = len(match) - 1
+			}
+		}
+
+		re := regexp.MustCompile(`\n[\t ]{` + fmt.Sprint(size) + `}`)
+		str = re.ReplaceAllString(str, "\n")
+	}
+
+	// 5. Remove leading whitespace.
+	re = regexp.MustCompile(`^\r?\n`)
+	str = re.ReplaceAllString(str, "")
+
+	return str
+}
+
+// checks if two strings are equal, treating any occurrences of `%%` in the
+// expected string to mean "any text"
+func areEqual(t *testing.T, actual, expect string) bool {
+	t.Helper()
+
+	expect = regexp.QuoteMeta(expect)
+	expect = strings.ReplaceAll(expect, "%%", ".+")
+
+	re := regexp.MustCompile(`^` + expect + `$`)
+
+	return re.MatchString(actual)
+}
+
+type cliTestCase struct {
+	name         string
+	args         []string
+	wantExitCode int
+	wantStdout   string
+	wantStderr   string
+}
+
+func testCli(t *testing.T, tc cliTestCase) {
+	t.Helper()
+
+	stdoutBuffer := &bytes.Buffer{}
+	stderrBuffer := &bytes.Buffer{}
+
+	ec := run(tc.args, stdoutBuffer, stderrBuffer)
+	// ec := run(tc.args, os.Stdout, os.Stderr)
+
+	stdout := stdoutBuffer.String()
+	stderr := stderrBuffer.String()
+
+	if ec != tc.wantExitCode {
+		t.Errorf("cli exited with code %d, not %d", ec, tc.wantExitCode)
+	}
+
+	if !areEqual(t, dedent(t, stdout), dedent(t, tc.wantStdout)) {
+		t.Errorf("stdout\n got:\n%s\n\n want:\n%s", dedent(t, stdout), dedent(t, tc.wantStdout))
+	}
+
+	if !areEqual(t, dedent(t, stderr), dedent(t, tc.wantStderr)) {
+		t.Errorf("stderr\n got:\n%s\n\n want:\n%s", dedent(t, stderr), dedent(t, tc.wantStderr))
+	}
+}
+
+func TestRun(t *testing.T) {
+	t.Parallel()
+
+	tests := []cliTestCase{
+		{
+			name:         "",
+			args:         []string{""},
+			wantExitCode: 1,
+			wantStdout: `
```

Comment:
Question about how maintainable this is, should I be running the scanner with the help flag manually and copying in the output if it looks correct?

Reply:
It should be very maintainable - you can just copy the result of the test suite:

```
--- FAIL: TestRun (0.00s)
    --- FAIL: TestRun/#00 (0.00s)
        main_test.go:239: stdout
             got:
            NAME:
               osv-scanner - scans various mediums for dependencies and matches it against the OSV database
            
            USAGE:
               osv-scanner [global options] command [command options] [directory1 directory2...]
            
            COMMANDS:
               help, h  Shows a list of commands or help for one command
            
            GLOBAL OPTIONS:
               --docker value, -D value [ --docker value, -D value ]      scan docker image with this name
               --lockfile value, -L value [ --lockfile value, -L value ]  scan package lockfile on this path
               --sbom value, -S value [ --sbom value, -S value ]          scan sbom file on this path
               --config value                                             set/override config file
               --json                                                     sets output to json (WIP) (default: false)
               --skip-git                                                 skip scanning git repositories (default: false)
               --recursive, -r                                            check subdirectories (default: true)
               --help, -h                                                 show help (default: false)
            
             want:
            NAME:
               osv-scanner - scans various mediums for dependencies and matches it against the OSV database
            
            USAGE:
               osv-scanner [global options] command [command options] [directory1 directory2...]
            
            COMMANDS:
               help, h  Shows a list of commands or help for one command
            
            GLOBAL OPTIONS:
               --docker value, -D value [ --docker value, -D value ]      scan docker image with this name
               --lockfile value, -L value [ --lockfile value, -L value ]  scan package lockfile on this path
               --sbom value, -S value [ --sbom value, -S value ]          scan sbom file on this path
               --config value                                             set/override config file
               --json                                                     sets output to json (WIP) (default: false)
               --skip-git                                                 skip scanning git repositories (default: false)
               --recursive, -r                                            check subdirectories (default: false)
               --help, -h                                                 show help (default: false)
    --- FAIL: TestRun/#02 (0.00s)
        main_test.go:239: stdout
             got:
            Scanning dir ./fixtures/locks-many/not-a-lockfile.toml
            NAME:
               osv-scanner - scans various mediums for dependencies and matches it against the OSV database
            
            USAGE:
               osv-scanner [global options] command [command options] [directory1 directory2...]
            
            COMMANDS:
               help, h  Shows a list of commands or help for one command
            
            GLOBAL OPTIONS:
               --docker value, -D value [ --docker value, -D value ]      scan docker image with this name
               --lockfile value, -L value [ --lockfile value, -L value ]  scan package lockfile on this path
               --sbom value, -S value [ --sbom value, -S value ]          scan sbom file on this path
               --config value                                             set/override config file
               --json                                                     sets output to json (WIP) (default: false)
               --skip-git                                                 skip scanning git repositories (default: false)
               --recursive, -r                                            check subdirectories (default: true)
               --help, -h                                                 show help (default: false)
            
             want:
            Scanning dir ./fixtures/locks-many/not-a-lockfile.toml
            NAME:
               osv-scanner - scans various mediums for dependencies and matches it against the OSV database
            
            USAGE:
               osv-scanner [global options] command [command options] [directory1 directory2...]
            
            COMMANDS:
               help, h  Shows a list of commands or help for one command
            
            GLOBAL OPTIONS:
               --docker value, -D value [ --docker value, -D value ]      scan docker image with this name
               --lockfile value, -L value [ --lockfile value, -L value ]  scan package lockfile on this path
               --sbom value, -S value [ --sbom value, -S value ]          scan sbom file on this path
               --config value                                             set/override config file
               --json                                                     sets output to json (WIP) (default: false)
               --skip-git                                                 skip scanning git repositories (default: false)
               --recursive, -r                                            check subdirectories (default: false)
               --help, -h                                                 show help (default: false)
    --- FAIL: TestRun/#05 (0.71s)
        main_test.go:239: stdout
             got:
            Scanning dir ./fixtures/locks-one-with-nested
            Scanned /home/gareth/workspace/projects-oss/osv-scanner/cmd/osv-scanner/fixtures/locks-one-with-nested/nested/composer.lock file and found 1 packages
            Scanned /home/gareth/workspace/projects-oss/osv-scanner/cmd/osv-scanner/fixtures/locks-one-with-nested/yarn.lock file and found 1 packages
            
             want:
            Scanning dir ./fixtures/locks-one-with-nested
            Scanned %%/fixtures/locks-one-with-nested/yarn.lock file and found 1 packages
FAIL
FAIL	github.com/google/osv-scanner/cmd/osv-scanner	0.907s
ok  	github.com/google/osv-scanner/internal/grouper	(cached)
?   	github.com/google/osv-scanner/internal/osv	[no test files]
?   	github.com/google/osv-scanner/internal/output	[no test files]
?   	github.com/google/osv-scanner/internal/sbom	[no test files]
ok  	github.com/google/osv-scanner/pkg/lockfile	(cached)
FAIL
```

----

It can be a pain to update at times when you've made a minor change to the output, but overall I've found it's far better having the peace of mind in being able to see the impact your change makes to all the different cli output situations (and most of the time you can do a bulk multi-caret copy and paste 😄)